### ISSUE TITLE
HG1604 Validation of 3.3.1604

### ISSUE THREAD

**Issue by jmarra13** [dmwm/CRABServer#5186]
The validation will be a bit diff from the last one, It will start with a stopped TW and few tasks to test old rest against a new TW.

**Comment by jmarra13** [comment]
Hi @mmascher the new TW is failing with this errors:

>[29/Mar/2016:13:43:46]  RESTSQL:lFpSxhIwnrYH execute: () {'tw_name': 'vocms045', 'get_status': 'QUEUED', 'limit': 10}
[29/Mar/2016:13:43:46]  SERVER REST ERROR WMCore.REST.Error.ExecutionError 9331b62d11d96f15226f2af87a5e8821 (Execution error)
[29/Mar/2016:13:43:46]    Traceback (most recent call last):
[29/Mar/2016:13:43:46]      File "/data/srv/beHG1603h/sw/slc6_amd64_gcc493/cms/crabserver/3.3.1603.rc1-comp/lib/python2.7/site-packages/WMCore/REST/Format.py", line 228, in stream_chunked
[29/Mar/2016:13:43:46]        for obj in stream:
[29/Mar/2016:13:43:46]      File "/data/srv/beHG1603h/sw/slc6_amd64_gcc493/cms/crabserver/3.3.1603.rc1-comp/lib/python2.7/site-packages/CRABInterface/RESTWorkerWorkflow.py", line 138, in get
[29/Mar/2016:13:43:46]        newtask.deserialize(row)
[29/Mar/2016:13:43:46]      File "/data/srv/beHG1603h/sw/slc6_amd64_gcc493/cms/crabserver/3.3.1603.rc1-comp/lib/python2.7/site-packages/CRABInterface/RESTWorkerWorkflow.py", line 241, in deserialize
[29/Mar/2016:13:43:46]        self['tm_scriptargs'] = literal_eval(task[42] if (task[42] is None or isinstance(task[42], str)) else task[42].read())
[29/Mar/2016:13:43:46]      File "/data/srv/beHG1603h/sw/slc6_amd64_gcc493/external/python/2.7.11/lib/python2.7/ast.py", line 80, in literal_eval
[29/Mar/2016:13:43:46]        return _convert(node_or_string)
[29/Mar/2016:13:43:46]      File "/data/srv/beHG1603h/sw/slc6_amd64_gcc493/external/python/2.7.11/lib/python2.7/ast.py", line 79, in _convert
[29/Mar/2016:13:43:46]        raise ValueError('malformed string')
[29/Mar/2016:13:43:46]    ValueError: malformed string
[29/Mar/2016:13:43:46]  SERVER HTTP ERROR cherrypy._cperror.HTTPError 1b1c62ed0248e7876563e1d4d35ddb0b ((500, 'Execution error'))
[29/Mar/2016:13:43:46]    Traceback (most recent call last):
[29/Mar/2016:13:43:46]      File "/data/srv/beHG1603h/sw/slc6_amd64_gcc493/cms/crabserver/3.3.1603.rc1-comp/lib/python2.7/site-packages/WMCore/REST/Server.py", line 697, in default
[29/Mar/2016:13:43:46]        return self._call(RESTArgs(list(args), kwargs))
[29/Mar/2016:13:43:46]      File "/data/srv/beHG1603h/sw/slc6_amd64_gcc493/cms/crabserver/3.3.1603.rc1-comp/lib/python2.7/site-packages/WMCore/REST/Server.py", line 807, in _call
[29/Mar/2016:13:43:46]        return stream_maybe_etag(apiobj.get('etag_limit', self.etag_limit), etagger, reply)
[29/Mar/2016:13:43:46]      File "/data/srv/beHG1603h/sw/slc6_amd64_gcc493/cms/crabserver/3.3.1603.rc1-comp/lib/python2.7/site-packages/WMCore/REST/Format.py", line 475, in stream_maybe_etag
[29/Mar/2016:13:43:46]        raise cherrypy.HTTPError(int(err), message)
[29/Mar/2016:13:43:46]    HTTPError: (500, 'Execution error')
[29/Mar/2016:13:43:46] vocms0132.cern.ch 128.142.142.23 "GET /crabserver/preprod/workflowdb?getstatus=QUEUED&limit=10&workername=vocms045 HTTP/1.1" 500 Internal Server Error [data: 3699 in 726 out 36136 us ] [auth: OK "/DC=ch/DC=cern/OU=computers/CN=tw/vocms052.cern.ch" "" ] [ref: "" "CRABClient/0.0.0" ]

**Comment by mmascher** [comment]
Did you run the changes on the db?
https://github.com/dmwm/CRABServer/blob/master/etc/updateOracle.sql#L72-L80

**Comment by jmarra13** [comment]
httplib2 package is missing at >/data/srv/TaskManager/3.3.1604.rc1/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1/lib/python2.7/site-packages/WMCore/Services/Requests.py

it was referenced in _getURLOpener but it is imported only inside makeRequest_httplib

Maybe we should move to another WMCore version which have it imported on the top and not inside of function definition.

**Comment by jmarra13** [comment]
TW is failing to update the logs right after the startup, maybe due to this error.

>[crab3@vocms045 current]$ cat nohup.out
Traceback (most recent call last):
  File "/data/srv/TaskManager/current/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1-comp/lib/python2.7/site-packages/TaskWorker/MasterWorker.py", line 306, in <module>
    mw.algorithm()
  File "/data/srv/TaskManager/current/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1-comp/lib/python2.7/site-packages/TaskWorker/MasterWorker.py", line 244, in algorithm
    worktype, failstatus = STATE_ACTIONS_MAP[task['tm_task_command']]
KeyError: None

**Comment by mmascher** [comment]
Hi @jmarra13 , it seems that the column "task command" for a task is None. The isntructions here https://github.com/dmwm/CRABServer/blob/master/etc/updateOracle.sql#L75-L80 should have taken care of this. Either I missed a case, or these instructions were not executed right before starting the TW.

**Comment by emaszs** [comment]
I had the same problem, and as Marco suggested, I was using an old version of the REST interface which wasn't providing the tm_task_command value in one of it's APIs. Updating it fixed the problem.

**Comment by jmarra13** [comment]
there is typo on masterworker.py, fixing locally for now.

>[crab3@vocms045 current]$ cat nohup.out 
Traceback (most recent call last):
  File "/data/srv/TaskManager/current/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1-comp/lib/python2.7/site-packages/TaskWorker/MasterWorker.py", line 306, in <module>
    mw.algorithm()
  File "/data/srv/TaskManager/current/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1-comp/lib/python2.7/site-packages/TaskWorker/MasterWorker.py", line 229, in algorithm
    self.failQueuedTasks()
  File "/data/srv/TaskManager/current/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1-comp/lib/python2.7/site-packages/TaskWorker/MasterWorker.py", line 213, in failQueuedTasks
    self.logger.debug("Failing QUEUED task %s", task['tm_task_name'])
KeyError: 'tm_task_name'
Traceback (most recent call last):
  File "/data/srv/TaskManager/current/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1-comp/lib/python2.7/site-packages/TaskWorker/MasterWorker.py", line 306, in <module>
    mw.algorithm()
  File "/data/srv/TaskManager/current/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1-comp/lib/python2.7/site-packages/TaskWorker/MasterWorker.py", line 229, in algorithm
    self.failQueuedTasks()
  File "/data/srv/TaskManager/current/slc6_amd64_gcc493/cms/crabtaskworker/3.3.1604.rc1-comp/lib/python2.7/site-packages/TaskWorker/MasterWorker.py", line 213, in failQueuedTasks
    self.logger.debug("Failing QUEUED task %s", task['tm_task_name'])
KeyError: 'tm_task_name'

@mmascher will fix on master branch.

**Comment by jmarra13** [comment]
dashboard shows wrong info for task[1], crab status[2] show jobs running or idle and dasboard show jobs in failed state


[1] http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=crab&refresh=0&table=Jobs&p=1&records=25&activemenu=0&status=&site=&tid=160331_094344%3Ajmsilva_crab_HG1604a-comp-4-Analysis_Use_Parent-L-T_O-T_P-T_IL-F

[2] 
jmsilva@lxplus069 ~/workspace/crabValidation/scripts> crab status /afs/cern.ch/work/j/jmsilva/crabValidation/HG1604a-comp/CMSSW_7_4_7/src/HG1604a-comp-4/crab_HG1604a-comp-4-Analysis_Use_Parent-L-T_O-T_P-T_IL-F --long
CRAB project directory:		/afs/cern.ch/work/j/jmsilva/crabValidation/HG1604a-comp/CMSSW_7_4_7/src/HG1604a-comp-4/crab_HG1604a-comp-4-Analysis_Use_Parent-L-T_O-T_P-T_IL-F
Task name:			160331_094344:jmsilva_crab_HG1604a-comp-4-Analysis_Use_Parent-L-T_O-T_P-T_IL-F
Grid scheduler:			crab3test-3@vocms096.cern.ch
Task status:			SUBMITTED
Dashboard monitoring URL:	http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=crab&refresh=0&table=Jobs&p=1&records=25&activemenu=2&status=&site=&tid=160331_094344%3Ajmsilva_crab_HG1604a-comp-4-Analysis_Use_Parent-L-T_O-T_P-T_IL-F
Warning:			The CRAB3 server backend detected lumis split across files in the input dataset. Will apply the necessary corrections in the splitting algorithms. You can ignore this message.

Jobs status:			idle           37.6% ( 38/101)
				running        62.4% ( 63/101)

**Comment by jmarra13** [comment]
Two config knobs were broke by WMCore version:

`useParent`

and

`secondaryInputDataset `

they work in prod with the same configuration used by validations tests but on preprod they run into a CMSSW error:

>== CMSSW: process id is 1694122 status is 65
======== CMSSW OUTPUT FINSHING ========
ERROR: Caught WMExecutionFailure - code = 65 - name = CmsRunFailure - detail = Error running cmsRun
{'arguments': ['/bin/bash', '/wntmp/condor/execute/dir_1464417/glide_mS7qJs/execute/dir_1693729/cmsRun-main.sh', '', 'slc6_amd64_gcc491', 'scramv1', 'CMSSW', 'CMSSW_7_4_7', 'FrameworkJobReport.xml', 'cmsRun', 'PSet.py', 'sandbox.tar.gz', '', '']}
Return code: 65
>
>NOTE: FJR has exit code 8001 and WMCore reports 65; preferring the FJR one.
ERROR: Exceptional exit at Thu Mar 31 10:31:54 2016 (8001): CmsRunFailure
CMSSW error message follows.
Fatal Exception
An exception of category 'PluginNotFound' occurred while
   [0] Constructing the EventProcessor
Exception Message:
Unable to find plugin 'CondorStatusService' in category 'CMS EDM Framework Service'. Please check spelling of name.

**Comment by mmascher** [comment]
Thanks @jmarra13 , maybe that's because it is an old version of CMSSW that does not have this CondorStatusService . @bbockelm warned us about this possibility.

Which CMSSW version are you using? I fear this might be wrong: https://github.com/dmwm/WMCore/commit/30583f4293592fb717ad46ecf54ebef38a5d6db5#diff8e00644fb812b3a8058f276f67b8b381R598

About the previous message (the dashboard thing), is this a reproducible error? I would like to exclude a temporary error on the dashboard side.

**Comment by belforte** [comment]
the dashboard thing is an URL constructed (badly) by CRAB itself. I do not think there's any reason to contact dashboard web site to create it.

**Comment by belforte** [comment]
about the 'Unable to find plugin CondorStatusService'.. how can it depend on the configuratin knob(s) ?

**Comment by mmascher** [comment]
Uhm.. we do not contact Dashboard to build the url. The CondorStatusService is a new CMSSW service that has to be added only for version of CMSSW that supports it. I see the job used CMSSW_7_4_7, apparently this version does not support the CondorStatusService, yet in the new WMCore the service is added (see `elif result.group(2) == 4 and result.group(3) >= 7:` in the link above)

**Comment by belforte** [comment]
as simple as changing `result.group(3) >= 7` to ` result.group(3) >= 8` ?
but let's get smething more reliable than "guess and try" on the miminum required CMSSW version number

**Comment by mmascher** [comment]
Yes, provided CMSSW_7_4_8 supports the CondorStatusService that should do it. In this thread  https://hypernews.cern.ch/HyperNews/CMS/get/wmDevelopment/737/1/2/1.html (well, the message after this one which does not have a hypernews url but I was cc-ed) @hufnagel said 7_4_7 supports it. Dirk, can you check this again please or tell us where to look for that?

**Comment by mmascher** [comment]
Also, I think we were lucky to catch this, should we add in our tests something to check different CMSSW releases?

**Comment by belforte** [comment]
condor plulging is here e.g.  https://github.com/cms-sw/cmssw/commits/CMSSW_8_1_X/FWCore/Services/plugins/CondorStatusUpdater.cc
but I am not capable to make GIT tell me which releases it was tagged into. Nor does http://cmslxr.fnal.gov/lxr/source/FWCore/Services/plugins/CondorStatusUpdater.cc help.
But I am sure there _must_ be a way to get this information from other then "words"

**Comment by bbockelm** [comment]
Hi,

Looked at the git tags - this feature started in 7_6_0.

Brian

**Comment by belforte** [comment]
CondorStatusService is available from CMSSW_7_6_0 onward. Only. See
https://hypernews.cern.ch/HyperNews/CMS/get/softwareDistrib/737/1/1.html

@jmarra13 can you tell me how to reproduce the 'dashboard thing' ?

**Comment by jmarra13** [comment]
@belforte the dataset used by those 2 templates needs CMSSW 7_4_7, acording to DAS[1], can I use a newer version? In this case CMSSW_7_6_0. I am trying to replicate the dashboard issue and will let you know.

[1] https://cmsweb.cern.ch/das/request?input=config%20dataset%3D/MinimumBias/Run2015B-PromptReco-v1/RECO&instance=prod/global&idx=0&limit=10

**Comment by mmascher** [comment]
The `CondorStatusService` error concerns the CMSSW release, not the fact that you used `useParent` and `secondaryInputDataset`. I spoke with @emaszs , on Monday he will take care of updating the crab WMCore branch (and will create a PR for the master as well)

**Comment by juztas** [comment]
@jmarra13 If I can suggest, just to keep up to date all the tests in CRAB3 branch.Also add new validation which takes like 5 different releases  (8_1_X, 8_0_X, 7_5_X, 7_4_X, 6_X_X) or do it in automatic way with scram to test always one newer release and submits a task for each with 5 jobs. It is far from ideal, but would be good, I think.
Later you can easily check dashboard and see the failures if there are any

**Comment by belforte** [comment]
I can read that dataset with CMSSW_8_0_0
In general I agree with Justas, let's always use a recent CMSSW version and change input datasets if needed. If you can (easily) test also previous releases at same time, so much the better, but it is uaually rare that we do something which breaks thing for old CMSSW

**Comment by jmarra13** [comment]
stageout fails due to a missing gfal-* commands, but in the log is written to use srmv2-lcg not gfal2.

>====== Mon Apr  4 13:28:17 2016: Starting initialization of stageout manager for local stageouts.
       -----> Stageout manager log start
StageOutMgr::\__init__()
==== Stageout configuration start ====
__Local Stage Out Implementation to be used is: srmv2-lcg__
Local Stage Out SE Name to be used is hepse01.colorado.edu
Local Stage Out PNN to be used is T3_US_Colorado
Local Stage Out Catalog to be used is trivialcatalog_file://nfs/hepcode/cmssw/cmssoft/cms/SITECONF/local/PhEDEx/storage.xml?protocol=srmv2
..
..
..
==== Stageout configuration finish ====
       <----- Stageout manager log finish
Initialization was ok.
====== Mon Apr  4 13:28:17 2016: Finished initialization of stageout manager for local stageouts (status 0).
Will use gfal2 commands for direct stageout.
====== Mon Apr  4 13:28:17 2016: Starting initialization of stageout implementation for direct stageouts.
Initialization was ok.
====== Mon Apr  4 13:28:17 2016: Finished initialization of stageout implementation for direct stageouts (status 0).
====== Mon Apr  4 13:28:17 2016: Starting remote stageout of user output files.
-----> Mon Apr  4 13:28:17 2016: Starting remote stageout of output.root.
       -----> Stageout implementation log start
2016-04-04T07:28:17 : Creating output directory...
2016-04-04T07:28:17 : Running the stage out...
__env: gfal-copy: No such file or directory__
gfal-copy exit status: 127

Same task has jobs that finish properly on sites where gfal commands were present.

**Comment by belforte** [comment]
I think that's unfortunately expected, the stagout was simply coded wrong from the beginning:
https://github.com/dmwm/CRABServer/issues/5156

**Comment by juztas** [comment]
@jmarra13 can you give me full log file ?

**Comment by juztas** [comment]
@jmarra13 you also want this: https://github.com/dmwm/WMCore/pull/6779
This is what you reported in the meeting

**Comment by jmarra13** [comment]
@juztas here is it https://cmsweb-testbed.cern.ch/scheddmon/059/cms488/160404_123954:jmsilva_crab_HG1604a-test_tag-1-Analysis_225Chars_WF-L-T_O-T_P-T_IL-F-DOC-T/job_out.6.0.txt

**Comment by jmarra13** [comment]
As reminded by @juztas even if task can run with newer CMSSW sometimes is needed to use an older version like this task[1], in this case the task needs CMSSW 5.3.22 and it fails to untar the sandbox in this version. A patch for this issue is here https://github.com/dmwm/WMCore/pull/6779


[1] http://dashb-cms-job.cern.ch/dashboard/templates/task-analysis/#user=Jadir+Marra+da+Silva&refresh=0&table=Jobs&p=1&records=-1&activemenu=0&status=&site=&tid=160404_125236%3Ajmsilva_crab_HG1604a-test_tag-1-PrivateMC_for_LHE-L-T_O-T_P-T_IL-F-DOC-T

**Comment by juztas** [comment]
Now is clear. This task or TW is configured to do first remote and do local if it fails to do remote.
Also you would want to check why it uses gfal-copy if command does not exist... It should not even try to do that. 
Now about local, remote, direct Andrew explained everything and it is not an issue which you get and it is worth to check his pull request.

**Comment by juztas** [comment]
for this CMSSW_5_3_22 it is needed for CRAB3, Checking releases we have all varieties of them and would not be nice to forbid users from using them...

http://dashb-cms-job.cern.ch/dashboard/templates/web-job2/#user=&refresh=0&table=Jobs&p=1&records=25&activemenu=1&usr=&site=&submissiontool=&application=&activity=analysis&status=&check=submitted&tier=&date1=2016-03-01+14%3A39&date2=2016-04-01+00%3A00&sortby=application&scale=linear&bars=20&ce=&rb=&grid=&jobtype=&submissionui=&dataset=&submissiontype=&task=&subtoolver=&genactivity=&outputse=&appexitcode=&accesstype=&inputse=&cores=

**Comment by jmarra13** [comment]
@juztas  TW is set to remote first, I did that due to several ASO timeouts in the beginning of the validation.

**Comment by juztas** [comment]
@jmarra13 I can see this from the log, but the point which I say is that there was a check if gfal-copy is available, use it, if not check if lcg-cp is available, if not fail. So in this case it used gfal-copy which is not available, there is no point to use it if it is not available and it should print warning.

**Comment by juztas** [comment]
I think because we check gfal-copy, but whenever stageout executes it executes with env -i ... gfal-copy. So just change the check to env gfal-copy and add some prints ;)

**Comment by belforte** [comment]
Justas, let's try to put some order, will  you ?
Are you talking about a problem introduced in some change in 3.3.1604 wrt to 3.3.1603 or simply 
adding fuel to the (very old) story that cmscp.py needs a good rewriting ?

**Comment by belforte** [comment]
It was made very clear (e.g. by Daniela Bauer) that env -i gfal-copy has no hope of working at every 
site. We can't start a war on sites by changing our code (exp. given that what we had at some point 
was working everyhwere). It was tried, it was discussed, I do not see it flying w/o much more effort 
than its worth. If really we have no solution, we have to spend the time to explain this very very 
well and very very clearly to all of our collegues, starting with Lammel/Bagliesi.

thanks

**Comment by juztas** [comment]
Stefano, I think there are more urgent things to be solved. About 'env -i' and without it we already saw huge failure on Purdue site with mixing environment. It was changed for production by mistake not to use env -i and started to fail a lot... Some links:
https://github.com/dmwm/WMCore/commit/68f64875567501bc83ee63fb6543f84319f85611
https://github.com/dmwm/WMCore/pull/6756
https://hypernews.cern.ch/HyperNews/CMS/get/computing-tools/1478/1/2/1.html
https://ggus.eu/index.php?mode=ticket_info&ticket_id=120327

I am not saying to start a war, but someone should come and say this is solution and it will work for all releases and etc... I don`t think we want to do ping-pong and one day add, another day remove it. I really don't see an easy solution for this env -i.... So far which is in my head is that we build own gfal-copy and maintain it in COMP environment. In this case sites are not forced to have it. But this requires longer discussions with Diego G. Workflow team, CRAB3 team...

I would do in this order for particular what Jadir found and to improve it a bit:
a) Check why it was using gfal-copy, while it is not available. it will be because we search for gfal-copy command but execution goes with env -i. So just add more prints and warnings in the logs. It is not changing how it works.
b) Another problem is what Andrew reported if site == dest_site and we use gfal-copy, we should use local stageout impl. So take a look at Andrew's pull request.
b) cmscp needs a rewriting, but it does its job correctly. If we want to make any change, it becomes a bit hard due to too many options. This one I think is long standing issue and sometimes it is better to write code in 'KISS' principle :)

**Comment by jmarra13** [comment]
crab status --sort=<runtime|memory|cpu> only works if you pass --json otherwise always return zeros or unknown values

**Comment by emaszs** [comment]
crab status --sort=cpu/memory/etc seems to work fine for me on preprod / my test vm, possibly something wrong with your client?

**Comment by bbockelm** [comment]
I think it's nuts to consider bootstrapping our own `gfal` (plus plugin code plus security libraries plus ...).

We're supposed to have a working `gfal` in the environment before CRAB3 is launched.  Stefano mentioned that restoring that environment 100% from launch has run into issues in the past - can you explain what the problem was then?  Why wouldn't restoring it here help?

I'm thinking about running `env -0 > original_env` in the CRAB3 bash script and altering the stageout plugin to load this and provide it to the `exec` syscall.

**Comment by mmascher** [comment]
Alternatively we can split cmscp in `cmscp_copy` and `cmscp_cmsweb`. The first is executed in whatever environment we find, the second one is executed after we source the COMP env

Having a look at the code it should not be too difficult (unless there are surprises). Maybe one day we'll not even need the cmscp_cmsweb (in principle now that postjobs are deferred we do not actually need to upload filemetadata and couch document in the job wrapper, but we can do it in the schedd)

**Comment by bbockelm** [comment]
Actually, I was thinking about having `cmscp` run in the same environment as it currently does - the original environment is passed as a dictionary to the stageout plugin and only the stageout script (i.e., the thing containing `gfal-copy`) gets the original environment.

**Comment by belforte** [comment]
Hey @bbockelm  though you were being chased by hungry iguanas now !
Anyhow.. I mentioned troubles with env save/restore out of vague CRAB2 memories from likely SL4 times. I too expect it to work, and maybe at the time the problem was that CRAB code (at least on client) could not restore the proper env. anyhow because crab was usually setup after scram had already messed up the env. Now we have scram unsetenv and life is easy. And surely for job wrapper we should be fine. in case my opinion matters, I like your way better than Marco's one. Let's remember to possibly pass X509_USER_PROXY to the stageout script.

**Comment by juztas** [comment]
mh.. Maybe like this:
This part: https://github.com/dmwm/CRABServer/blob/master/scripts/gWMS-CMSRunAnalysis.sh#L147-L171 and this part https://github.com/dmwm/WMCore/blob/master/etc/submit.sh#L25-L43 are equal and this is what we need for gfal-copy. 
      1) create a separate bash script and in CRAB3/Prod use that script in bootstrap before the job;
      2) change gfal2 plugin in WMCore and do:
Now:
```
env -i X509_USER_PROXY=$X509_USER_PROXY sh -c 'printenv; date; gfal-rm -vvv -t 600 /store/temp/user....'
```
New
```
env -i X509_USER_PROXY=$X509_USER_PROXY sh -c 'source grid_env.sh; printenv; date; gfal-rm -vvv -t 600 /store/temp/user....'
```
I think this would solve it for all sites and also what Daniela reported, no?
Another thing is that CRAB3 is not yet near master branch of WMCore, and now it is a bit different how crab and wma is using it, but it is just double pull request creation :)

**Comment by belforte** [comment]
what exactly would grid_env.sh ?

**Comment by juztas** [comment]
this part:  https://github.com/dmwm/CRABServer/blob/master/scripts/gWMS-CMSRunAnalysis.sh#L147-L171 or same which is in WMCore for prod jobs.

**Comment by bbockelm** [comment]
@juztas - unfortunately, this is how you would do it on EGI grid sites but not OSG.  GlideinWMS already handles these differences - that's why I suggested preserving the environment.

**Comment by belforte** [comment]
@juztas  https://github.com/dmwm/CRABServer/blob/master/scripts/gWMS-CMSRunAnalysis.sh#L147-L171 will only restore the CMS specific part of the environment, but placing grid client in the env is done before that, care of the site admin (or glidein) and there's no guaranteed way to do that on our own. We've been there. For a few years, accidentally, COMP env has been compatible with voms-proxy-* and lcg-* env, and we relaxed. Time to be very strict and hopefully will be OK for the forseable future.

**Comment by juztas** [comment]
you are both right. I was not aware of this. Maybe did not understood proposed solution, but why to pass it as a dictionary and do it in python? Maybe it is simplier before job startup to save all environment in bash tofile `startup_env.sh` and before doing gfal-copy source this file. I am just thinking how to make it easier and in case in future any other command needs this clean wn environment, it can source it. Also this would be minimal change for Prod and CRAB3, I think

For this specific CMS part, it is additional feature, but it would be easier to maintain it in one place, liek WMCore, than now in 3 :)
